src/sql/src/plan/query.rs

```diff
@@ -701,36 +702,71 @@ pub fn plan_mutation_query(
         bail!("cannot mutate system table '{}'", table.name());
     }
 
-    let get = HirRelationExpr::Get {
+    let mut relation_expr = HirRelationExpr::Get {
         id: Id::Global(table.id()),
         typ: desc.typ().clone(),
     };
 
-    let selection: HirRelationExpr = match selection {
-        Some(mut expr) => {
-            transform_ast::transform_expr(scx, &mut expr)?;
-            let expr = resolve_names_expr(&mut qcx, expr)?;
-            let ecx = &ExprContext {
-                qcx: &qcx,
-                name: "WHERE clause",
-                scope: &Scope::from_source(
-                    Some(PartialName::from(table.name().clone())),
-                    desc.iter_names(),
-                    None,
-                ),
-                relation_type: &RelationType::new(desc.iter_types().cloned().collect()),
-                allow_aggregates: false,
-                allow_subqueries: true,
+    let relation_type = qcx.relation_type(&relation_expr);
+
+    let scope = Scope::from_source(
+        Some(PartialName::from(table.name().clone())),
+        desc.iter_names(),
+        None,
+    );
+
+    if let Some(mut expr) = selection {
+        // If `DELETE FROM <table_name> USING <using> WHERE <selection>`...
+        if !using.is_empty() {
+            // Convert `using` + `selection` to existential subquery, which can
+            // be correlated to the outer scope, i.e. rewrite to `EXISTS (SELECT
+            // 1 FROM <using> WHERE <selection>)`.
+            let subselect = Select {
+                distinct: None,
+                // From pg: Since the result [of EXISTS (<subquery>)] depends
+                // only on whether any rows are returned, and not on the
+                // contents of those rows, the output list of the subquery is
+                // normally unimportant. A common coding convention is to write
+                // all EXISTS tests in the form EXISTS(SELECT 1 WHERE ...)
+                // https://www.postgresql.org/docs/14/functions-subquery.html
+                projection: vec![SelectItem::Expr {
+                    expr: Expr::Value(Value::Number("1".into())),
+                    alias: None,
+                }],
+                from: using,
+                selection: Some(expr),
```

Comment:
I think we still need to generate the subquery even when there is not `WHERE` clause, since the DELETE should be a no-op if the join after the USING clause doesn't return any row, like postgres does:

```
asenac=> create table t1(f1 int);
CREATE TABLE
asenac=> insert into t1 values (1);
INSERT 0 1
asenac=> create table t2(f1 int);
CREATE TABLE
asenac=> select * from t1;
 f1 
----
  1
(1 row)

asenac=> delete from t1 using t2;
DELETE 0
asenac=> select * from t1;
 f1 
----
  1
(1 row)

asenac=> 
```